src/VisualStudio/Core/Def/Implementation/Watson/WatsonReporter.cs

```diff
@@ -95,7 +112,12 @@ public static void Report(string description, Exception exception, Func<IFaultUt
             // watson and telemetry. 
             faultEvent.SetExtraParameters(exception, emptyCallstack);
 
-            TelemetryService.DefaultSession.PostEvent(faultEvent);
+            s_asyncTasks.Add(ThreadHelper.JoinableTaskFactory.RunAsync(async () =>
```

Comment:
This won't add the `JoinableTask` to the `JoinableTaskCollection` until after the first yielding await (when `RunAsync` actually returns to the caller). If you were already on a background thread when you called this, the only await in that delegate would not return and therefore you wouldn't have the protection you thought you had.
You should create a `JoinableTaskFactory s_asyncTasksFactory` field next to your `JoinableTaskCollection`, and initialize the factory based on the collection (the factory constructor takes a collection as an optional argument). Then use that factory here. That will guarantee that the JoinableTask is added to the collection even before invoking the delegate.

Be careful with static field initializers depending on one another, of course. I would be explicit about it by using a static cctor:

```cs
static WatsonReporter()
{
  s_asyncTasks = ThreadHelper.JoinableTaskContext.CreateCollection();
  s_asyncTasksFactory = ThreadHelper.JoinableTaskContext.CreateFactory(s_asyncTasks);
}
```